ios-base/Networking/Services/BaseAPIClient.swift

```diff
@@ -0,0 +1,131 @@
+import Foundation
+
+/// Concrete implementation of API Client.
+internal final class BaseAPIClient: APIClient {
+
+  // MARK: - Properties
+
+  private let emptyDataStatusCodes: Set<Int> = [204, 205]
+
+  private(set) var encodingConfiguration: EncodingConfiguration = .default
+
+  private(set) var decodingConfiguration: DecodingConfiguration = .default
+
+  private(set) var headersProvider: HeadersProvider = RailsAPIHeadersProvider(
+    sessionHeadersProvider: SessionHeadersProvider()
+  )
+
+  private let networkProvider: NetworkProvider
+
+  required init(networkProvider: NetworkProvider) {
+    self.networkProvider = networkProvider
+  }
+
+  convenience init(
+    networkProvider: NetworkProvider,
+    headersProvider: HeadersProvider,
+    encodingConfiguration: EncodingConfiguration = .default,
+    decodingConfiguration: DecodingConfiguration = .default
+  ) {
+    self.init(networkProvider: networkProvider)
+    self.encodingConfiguration = encodingConfiguration
+    self.decodingConfiguration = decodingConfiguration
+    self.headersProvider = headersProvider
+  }
+
+  @discardableResult
+  func request<T: Decodable>(
+    endpoint: Endpoint,
```

Comment:
How are clients suppose to override headers? 
Shouldn't this function also receive a headersProvider parameter? Otherwise directly receive an APIEndpoint?

Reply:
the `Endpoint` protocol already provides a way to send headers with:
`var headers: [String: String] { get }`

In case you need custom headers for your API(which is the case of the BaseAPIClient for Rails.), we use inject  a HeadersProvider conformant object. The API Client is the one responsible to determine how and which headers to send.

Reply:
got it! please see my next question